### ISSUE TITLE
Slow sum/mean/max/... matrix operations compared to PyTorch

### ISSUE THREAD

**Issue by arnaudvl** [cupy/cupy#4896]
`x.sum()` where `x` is an NxN matrix seems to be significantly slower than e.g. the PyTorch equivalent (87x slower for N=10,000). This behaviour is also observed for other reduce operations such as `.mean()` or `.max()` and gets worse as N grows.

* Conditions (you can just paste the output of `python -c 'import cupy; cupy.show_config()'`)

```
OS                           : Linux-5.4.0-42-generic-x86_64-with-debian-buster-sid
CuPy Version                 : 8.5.0
NumPy Version                : 1.19.2
SciPy Version                : 1.6.1
Cython Build Version         : 0.29.22
CUDA Root                    : /usr
CUDA Build Version           : 10000
CUDA Driver Version          : 11000
CUDA Runtime Version         : 10000
cuBLAS Version               : 10000
cuFFT Version                : 10000
cuRAND Version               : 10000
cuSOLVER Version             : (10, 0, 0)
cuSPARSE Version             : 10000
NVRTC Version                : (10, 0)
Thrust Version               : 100903
CUB Build Version            : 100800
cuDNN Build Version          : 7605
cuDNN Version                : 7605
NCCL Build Version           : 2604
NCCL Runtime Version         : 2604
cuTENSOR Version             : None
Device 0 Name                : GeForce RTX 2080 Ti
Device 0 Compute Capability  : 75
Device 1 Name                : GeForce RTX 2080 Ti
Device 1 Compute Capability  : 75
Device 2 Name                : GeForce RTX 2080 Ti
Device 2 Compute Capability  : 75
Device 3 Name                : GeForce RTX 2080 Ti
Device 3 Compute Capability  : 75
```
PyTorch Version: 1.8.0

* Code to reproduce

```python
import cupy as cp
import numpy as np
from timeit import default_timer as timer
import torch


dim = [100, 1000, 5000, 10000]
n_runs = 10
n_warmup = 2
n_tot = n_runs + n_warmup

# place on device
device = torch.device('cuda')

# timers
t_start = torch.cuda.Event(enable_timing=True)
t_end = torch.cuda.Event(enable_timing=True)

# PyTorch
for d in dim:
    t_pt = np.zeros(n_tot)
    for n in range(n_tot):
        np.random.seed(0)
        x = np.random.randn(*(d, d)).astype(np.float32)
        x_pt = torch.from_numpy(x).to(device)
        torch.cuda.synchronize()
        t_start.record()
        x_pt.sum()
        t_end.record()
        torch.cuda.synchronize()
        t_pt[n] = 1e-3 * t_start.elapsed_time(t_end)  # 1e-3 b/c prints in ms
    t_mean, t_std = t_pt[n_warmup:].mean(), t_pt[n_warmup].std()
    print(f'PyTorch: {d} x {d}: {t_mean:.5f} +- {t_std:.5f}')

# CuPy
for d in dim:
    t_cp = np.zeros(n_tot)
    for n in range(n_tot):
        np.random.seed(0)
        x = np.random.randn(*(d, d)).astype(np.float32)
        x_cp = cp.asarray(x)
        cp.cuda.Stream.null.synchronize()
        t_start = timer()
        x_cp.sum()
        cp.cuda.Stream.null.synchronize()
        t_cp[n] = timer() - t_start
    t_mean, t_std = t_cp[n_warmup:].mean(), t_cp[n_warmup].std()
    print(f'CuPy: {d} x {d}: {t_mean:.5f} +- {t_std:.5f}')
```
Code is run on a single RTX2080Ti by setting `CUDA_VISIBLE_DEVICES={device nb}`

* Error messages, stack traces, or logs

Mean run time (in sec) and standard deviation over 10 runs after the warmup phase:

```
PyTorch: 100 x 100: 0.00002 +- 0.00000
PyTorch: 1000 x 1000: 0.00004 +- 0.00000
PyTorch: 5000 x 5000: 0.00024 +- 0.00000
PyTorch: 10000 x 10000: 0.00075 +- 0.00000
CuPy: 100 x 100: 0.00003 +- 0.00000
CuPy: 1000 x 1000: 0.00042 +- 0.00000
CuPy: 5000 x 5000: 0.01645 +- 0.00000
CuPy: 10000 x 10000: 0.06552 +- 0.00000
```

**Comment by emcastillo** [comment]
Try using CUB or cuTENSOR with the `CUPY_ACCELERATORS` env var.

https://docs.cupy.dev/en/stable/reference/environment.html
